### ISSUE TITLE
Handling non-blocking file descriptors consistently.

### ISSUE THREAD

**Issue by ioquatix** [axboe/liburing#364]
This is more of a discussion than an actual issue at the moment, but I hope to provide more concrete details as we explore the use cases.

We have an interesting problem regarding `IORING_OP_READV`/`IORING_OP_WRITEV`. In a general programming language environment, we don't always know if the file descriptor is in non-blocking mode or not. 

But the behaviour of `IORING_OP_READV`/`IORING_OP_WRITEV` depends on it. In non-blocking mode, we can get a result of `EAGAIN`.

However, this is not very useful in the case of even driven concurrency where we want operations like this to be "asynchronous" "blocking" - i.e. the CQE should not be generated until the read/write could make progress.

In general, my understanding is, we'd need 3-4 system calls to work around this, something like this:

```
int current_state = get_nonblocking_state(fd);
if (current_state != nonblocking) set_nonblocking_state(fd);
io_uring_read(..., fd, ...)
if (nonblocking_state_was_changed) restore_nonblocking_state(fd, current_state);
```

We can cache this to reduce the round trips, but it feels like this would be something better solved by a flag on `IORING_OP_READV`/`IORING_OP_WRITEV`. In fact, I'm having a hard time imagining any valid use case for non-blocking operation in the uring.

I'll try to test out specific cases where this is a problem and report back.

**Comment by YoSTEALTH** [comment]
Whole point of `io_uring` is its asynchronous by design! If there is a blocking call it either handles in async or opens a new kernel thread for that request.

By using `O_NONBLOCK` what you are saying is, let me choose what to do next when blocking happens vs letting `io_uring` deal with it.

There already is an `flags` option.

``` Python
fd = open('path/to/file', O_RDONLY)  # normal blocking read only open.
...
io_uring_prep_readv(sqe, fd, iovecs, nr_vecs, offset)
sqe.rw_flags |= RWF_NOWAIT  # now read is non-blocking
...
# catch `-EGAIN` in `cqe.res`, ...
```

**Comment by ioquatix** [comment]
Is there a flag for the opposite of `RWF_NOWAIT`?

**Comment by YoSTEALTH** [comment]
Not that I know of. If you want a blocking read just use normal `read()` don't use `io_uring_read*`

**Comment by ioquatix** [comment]
We have file descriptors that might already have O_NONBLOCK set. But we still want to use them with io_uring without getting EAGAIN.

**Comment by YoSTEALTH** [comment]
Hmm that's a tricky situation. Monkey patching `O_NONBLOCK` flag to not be non-blocking might lead to other problems/bugs.

If you have access to when the original flag is being set

`original_flags = O_CREAT | O_RDWR | O_NONBLOCK`

You could add a check like
``` python
if original_flags & O_NONBLOCK:
    # raise error saying `O_NONBLOCK` is not supported, invalid or something!
```
I am not really proud of that solution.

Why not just catch `-EGAIN` and poll it? `io_uring_prep_poll_add(sqe, fd, POLLIN)`

**Comment by ioquatix** [comment]
Yes it’s one option but we wanted to use the direct read/write operations if possible.

**Comment by YoSTEALTH** [comment]
suppose you can also remove the flag by:
``` Python
original_flags = O_CREAT | O_RDWR | O_NONBLOCK`
new_flags = original_flags & ~O_NONBLOCK
```

**Comment by axboe** [comment]
It would be a pretty trivial change to arm poll rather than return -EAGAIN to the application, if O_NONBLOCK is set on the file. The main issue here is how to inform io_uring of that desired behavior. OP_READV/OP_WRITEV flags are really readv2/writev2 flags, so it'd then be a generic change to allow this. Which arguably would make sense, as the justification would be that you can do blocking IO without having the query/set new file flags.

This applies to more than OP_READV/OP_WRITEV, though. It also applies to SEND/RECV and friends, so it's actually more of a generic change. Which means that an IOSQE_ flag would make more sense. The only potential issue there is that the flag space is pretty small, we have 8 bits and 6 are already used. Not a huge issue, we have two bits left, and the last bit could always be turned into a "look at extended bits" or something like that.

In any case, let me know if you want to test a patch for this. Should be easy to cook up.

**Comment by ioquatix** [comment]
I am happy to test and help with this.

**Comment by ioquatix** [comment]
Just thinking through what you described, do you think it would make sense for us to propose a flag to readv2/writev2 to be more specific about this?

That way, this doesn't become an io_uring specific feature, but works at the kernel level in a generic way.

**Comment by axboe** [comment]
As mentioned, that only then works for the read/write variants, not for any other opcode. Hence I'd be leaning towards making it an sqe flag instead. If it was just applicable to OP_READV and friends, then the RWF approach would be preferable.

**Comment by ioquatix** [comment]
Do other opcodes also take flags? e.g. accept, open, close, etc.

**Comment by axboe** [comment]
Each opcode has a specific set of flags, like the RWF_* for read/write, but there's also a generic set of IOSQE_* flags that apply to any opcode. That handles things like links, etc. Things that sit above the specific request type.

**Comment by ioquatix** [comment]
Oh that makes total sense. Okay, I'm keen to test this when you are ready.

cc @ciconia

**Comment by axboe** [comment]
I’ll spin a patch tomorrow morning.

**Comment by YoSTEALTH** [comment]
> It would be a pretty trivial change to arm poll rather than return -EAGAIN to the application, if O_NONBLOCK is set on the file.

@axboe O.o something like `IOSQE_IO_BLOCKING_POLL*`? If you can make this into an feature I would totally welcome it as well. Currently I have lines of code just checking `-EAGAIN` and polling it to account for both blocking/non-blocking flags. This feature would save many unnecessary round-trips.

Rather then just account for `EAGAIN` it would be nice if it accounted for "BLOCKING" states like `EAGAIN, EALREADY, EINPROGRESS, EWOULDBLOCK, ...`

**Comment by ioquatix** [comment]
I agree, the vast majority of user space code doesn't care about `EAGAIN`, it's like the half way house of concurrency.

**Comment by ciconia** [comment]
That's a great idea, adding such a flag will be very useful! Keep in mind though that if you want to support older Linux kernel versions (once this feature lands in a kernel release), you'll still have to provide some workaround, either resetting `O_NONBLOCK` or handling `EAGAIN`.

**Comment by axboe** [comment]
Probably something like IOSQE_IGNORE_NONBLOCK or whatever. It'll just be a way to force the normal poll path for pollable IO even if the fd is marked non-blocking.

**Comment by axboe** [comment]
Here's a first attempt. Let me know if you'd prefer a patch somewhere, not sure what this does to formatting... If it destroys whitespace, probably just use patch -l to ignore that part.

```
diff --git a/fs/io_uring.c b/fs/io_uring.c
index 46a25a7cb70a..25f72f35902d 100644
--- a/fs/io_uring.c
+++ b/fs/io_uring.c
@@ -108,7 +108,7 @@
 
 #define SQE_VALID_FLAGS	(IOSQE_FIXED_FILE|IOSQE_IO_DRAIN|IOSQE_IO_LINK|	\
 				IOSQE_IO_HARDLINK | IOSQE_ASYNC | \
-				IOSQE_BUFFER_SELECT)
+				IOSQE_BUFFER_SELECT | IOSQE_IGNORE_NONBLOCK)
 
 #define IO_TCTX_REFS_CACHE_NR	(1U << 10)
 
@@ -704,6 +704,7 @@ enum {
 	REQ_F_HARDLINK_BIT	= IOSQE_IO_HARDLINK_BIT,
 	REQ_F_FORCE_ASYNC_BIT	= IOSQE_ASYNC_BIT,
 	REQ_F_BUFFER_SELECT_BIT	= IOSQE_BUFFER_SELECT_BIT,
+	REQ_F_IGNORE_NONBLOCK_BIT	= IOSQE_IGNORE_NONBLOCK_BIT,
 
 	/* first byte is taken by user flags, shift it to not overlap */
 	REQ_F_FAIL_BIT		= 8,
@@ -740,6 +741,8 @@ enum {
 	REQ_F_FORCE_ASYNC	= BIT(REQ_F_FORCE_ASYNC_BIT),
 	/* IOSQE_BUFFER_SELECT */
 	REQ_F_BUFFER_SELECT	= BIT(REQ_F_BUFFER_SELECT_BIT),
+	/* IOSQE_IGNORE_NONBLOCK */
+	REQ_F_IGNORE_NONBLOCK	= BIT(REQ_F_IGNORE_NONBLOCK_BIT),
 
 	/* fail rest of links */
 	REQ_F_FAIL		= BIT(REQ_F_FAIL_BIT),
@@ -2444,6 +2447,15 @@ static bool io_resubmit_prep(struct io_kiocb *req)
 	return true;
 }
 
+/*
+ * Honor REQ_F_NOWAIT only if REQ_F_IGNORE_NONBLOCK isn't set
+ */
+static bool io_req_nowait(struct io_kiocb *req)
+{
+	return (req->flags & (REQ_F_NOWAIT | REQ_F_IGNORE_NONBLOCK)) ==
+		REQ_F_NOWAIT;
+}
+
 static bool io_rw_should_reissue(struct io_kiocb *req)
 {
 	umode_t mode = file_inode(req->file)->i_mode;
@@ -2451,7 +2463,7 @@ static bool io_rw_should_reissue(struct io_kiocb *req)
 
 	if (!S_ISBLK(mode) && !S_ISREG(mode))
 		return false;
-	if ((req->flags & REQ_F_NOWAIT) || (io_wq_current_is_worker() &&
+	if (io_req_nowait(req) || (io_wq_current_is_worker() &&
 	    !(ctx->flags & IORING_SETUP_IOPOLL)))
 		return false;
 	/*
@@ -3223,7 +3235,7 @@ static bool io_rw_should_retry(struct io_kiocb *req)
 	struct kiocb *kiocb = &req->rw.kiocb;
 
 	/* never retry for NOWAIT, we just complete with -EAGAIN */
-	if (req->flags & REQ_F_NOWAIT)
+	if (io_req_nowait(req))
 		return false;
 
 	/* Only for buffered IO */
@@ -3303,7 +3315,7 @@ static int io_read(struct io_kiocb *req, unsigned int issue_flags)
 		if (!force_nonblock && !(req->ctx->flags & IORING_SETUP_IOPOLL))
 			goto done;
 		/* no retry on NONBLOCK nor RWF_NOWAIT */
-		if (req->flags & REQ_F_NOWAIT)
+		if (io_req_nowait(req))
 			goto done;
 		/* some cases will consume bytes even on error returns */
 		iov_iter_revert(iter, io_size - iov_iter_count(iter));
@@ -3311,7 +3323,7 @@ static int io_read(struct io_kiocb *req, unsigned int issue_flags)
 	} else if (ret == -EIOCBQUEUED) {
 		goto out_free;
 	} else if (ret <= 0 || ret == io_size || !force_nonblock ||
-		   (req->flags & REQ_F_NOWAIT) || !(req->flags & REQ_F_ISREG)) {
+		   io_req_nowait(req) || !(req->flags & REQ_F_ISREG)) {
 		/* read all, failed, already did sync or don't want to retry */
 		goto done;
 	}
@@ -3434,7 +3446,7 @@ static int io_write(struct io_kiocb *req, unsigned int issue_flags)
 	if (ret2 == -EOPNOTSUPP && (kiocb->ki_flags & IOCB_NOWAIT))
 		ret2 = -EAGAIN;
 	/* no retry on NONBLOCK nor RWF_NOWAIT */
-	if (ret2 == -EAGAIN && (req->flags & REQ_F_NOWAIT))
+	if (ret2 == -EAGAIN && io_req_nowait(req))
 		goto done;
 	if (!force_nonblock || ret2 != -EAGAIN) {
 		/* IOPOLL retry should happen for io-wq threads */
@@ -6455,7 +6467,7 @@ static void __io_queue_sqe(struct io_kiocb *req)
 		} else {
 			io_put_req(req);
 		}
-	} else if (ret == -EAGAIN && !(req->flags & REQ_F_NOWAIT)) {
+	} else if (ret == -EAGAIN && !io_req_nowait(req)) {
 		if (!io_arm_poll_handler(req)) {
 			/*
 			 * Queued up for async execution, worker will release
diff --git a/include/uapi/linux/io_uring.h b/include/uapi/linux/io_uring.h
index f1f9ac114b51..582eee6e898b 100644
--- a/include/uapi/linux/io_uring.h
+++ b/include/uapi/linux/io_uring.h
@@ -70,6 +70,7 @@ enum {
 	IOSQE_IO_HARDLINK_BIT,
 	IOSQE_ASYNC_BIT,
 	IOSQE_BUFFER_SELECT_BIT,
+	IOSQE_IGNORE_NONBLOCK_BIT,
 };
 
 /*
@@ -87,6 +88,8 @@ enum {
 #define IOSQE_ASYNC		(1U << IOSQE_ASYNC_BIT)
 /* select buffer from sqe->buf_group */
 #define IOSQE_BUFFER_SELECT	(1U << IOSQE_BUFFER_SELECT_BIT)
+/* ignore if file has O_NONBLOCK set */
+#define IOSQE_IGNORE_NONBLOCK	(1U << IOSQE_IGNORE_NONBLOCK_BIT)
 
 /*
  * io_uring_setup() flags
```

**Comment by ioquatix** [comment]
I will try this out over the weekend.

**Comment by axboe** [comment]
> I will try this out over the weekend.

I'll run some quick testing on this later, I'll post an updated version here if needed.

**Comment by YoSTEALTH** [comment]
That's neat,

Does it account for cases where `O_NONBLOCK` is not set but `MSG_DONTWAIT` flag is set directly into socket `*recv`?

What about `*openat2` with `RESOLVE_CACHED` flag where `fd` hasn't been created yet but you are waiting for one.

Can this be in next Linux `5.13-rc*` patch or its more of a `5.14` thing.

**Comment by ioquatix** [comment]
I'm hoping we can transparently support this by using the poll fallback - i.e. assume it's going to work, but if we get EAGAIN, go to a fallback path.

**Comment by isilence** [comment]
Ok, seems I need to ask unpopular questions...

Disregarding generality for other io_uring requests, I have a problem with that design. We have a behaviour, blocking or not, and it is specified by rw flags (and also by file flags). So, it's blocking  by default and `RWF_NOWAIT` overwrites the behaviour to non-blocking. This one will be overriding overridden behaviour, that doesn't sound good from the design perspective.
To give the idea about concern, _reductio to absurdium_ here would be to also add `IOSQE_IGNORE_IGNORE_NONBLOCK`

So, the question here, does it add something that can't be currently achieved? (e.g. by clearing/not non-blocking flags). Finer blocking control or just yet another BLOCK/NONBLOCK flag doing the same thing? If there are inconsistencies in io_uring handling noblock (and I remember there were at some point), we need to figure out and document how it should be treated.

Also curious, assuming that the library takes fds from the user but doesn't take over the ownership, i.e. the user may use it in any other way it wish in parallel:
1) How come that you force 'blocking' behaviour for userspace passing in non-blocking file, when it can be the desired outcome?
2) Why the requirement of passing non-blocking files can't be forced on the user?

**Comment by isilence** [comment]
> Can this be in next Linux `5.13-rc*` patch or its more of a `5.14` thing.

@YoSTEALTH, 5.13 is feature closed and will be released pretty soon, so definitely not 5.13

**Comment by axboe** [comment]
Pavel, outside of the patch being a quick'n dirty, the intent is _only_ to ignore O_NONBLOCK if set. We should not be fiddling with RWF_NOWAIT or ditto send/recvmsg flags. Just for per-file state, not per request.

**Comment by YoSTEALTH** [comment]
@axboe My initial thought was `if cqe.res == -EAGAIN and IOSQE_new_flag:` wait/poll for task to finish at kernel-space, thus preventing a round trip to user-space and back. This way no set flags are messed with only checks for `cqe` end of the data. Is that even close to whats happening?

**Comment by axboe** [comment]
What you described, but only if O_NONBLOCK is set on the file. If you are setting request specific flags, like RWF_NOWAIT, then the -EAGAIN is returned as before. At that point you are specifically asking for non-block checking, the new sqe flag should only apply to state that you don't necessarily control (eg the file flags).

**Comment by YoSTEALTH** [comment]
@axboe OK, thanks you for clarifying. So it can only be used in `*open*` atm. Which is such a small use-case! wouldn't it have been better for developer to `new_flags = original_flags & ~O_NONBLOCK` way to remove the flag.

As much as I would like to see new feature added to `io_uring` I also wouldn't want to waste resources that could be better used!

**Comment by axboe** [comment]
No, I think you are misunderstanding. For requests that have a _specific_ non-block flag, like RWF_NOWAIT for read/write or MSG_NOWAIT for send/recv, it doesn't apply to those flags. The logic here is that you can just not set those flags, it's pointless to have an sqe flag that says ignore the flag I explicitly set for non-blocking data/space checking. It _only_ applies to O_NONBLOCK, as you don't necessarily control that flag or have exclusive ownership over the fd.

Hence it applies to any request type that makes use of O_NONBLOCK. And I believe that is what you are saying in the next sentence. If IOSQE_IGNORE_NONBLOCK is set, we don't return -EAGAIN just because O_NONBLOCK is set. We arm poll.

**Comment by YoSTEALTH** [comment]
@axboe OK, this applies after `open` is already created and you are stuck with `fd` with a `O_NONBLOCK` set on it, gotcha. As always you are thinking 10 steps ahead

**Comment by axboe** [comment]
Here's a simpler variant. Going to test this now.

```
diff --git a/fs/io_uring.c b/fs/io_uring.c
index fc8637f591a6..2d42273bb50f 100644
--- a/fs/io_uring.c
+++ b/fs/io_uring.c
@@ -108,7 +108,7 @@
 
 #define SQE_VALID_FLAGS	(IOSQE_FIXED_FILE|IOSQE_IO_DRAIN|IOSQE_IO_LINK|	\
 				IOSQE_IO_HARDLINK | IOSQE_ASYNC | \
-				IOSQE_BUFFER_SELECT)
+				IOSQE_BUFFER_SELECT | IOSQE_IGNORE_NONBLOCK)
 #define IO_REQ_CLEAN_FLAGS (REQ_F_BUFFER_SELECTED | REQ_F_NEED_CLEANUP | \
 				REQ_F_POLLED | REQ_F_INFLIGHT | REQ_F_CREDS)
 
@@ -706,6 +706,7 @@ enum {
 	REQ_F_HARDLINK_BIT	= IOSQE_IO_HARDLINK_BIT,
 	REQ_F_FORCE_ASYNC_BIT	= IOSQE_ASYNC_BIT,
 	REQ_F_BUFFER_SELECT_BIT	= IOSQE_BUFFER_SELECT_BIT,
+	REQ_F_IGNORE_NONBLOCK_BIT	= IOSQE_IGNORE_NONBLOCK_BIT,
 
 	/* first byte is taken by user flags, shift it to not overlap */
 	REQ_F_FAIL_BIT		= 8,
@@ -743,6 +744,8 @@ enum {
 	REQ_F_FORCE_ASYNC	= BIT(REQ_F_FORCE_ASYNC_BIT),
 	/* IOSQE_BUFFER_SELECT */
 	REQ_F_BUFFER_SELECT	= BIT(REQ_F_BUFFER_SELECT_BIT),
+	/* IOSQE_IGNORE_NONBLOCK */
+	REQ_F_IGNORE_NONBLOCK	= BIT(REQ_F_IGNORE_NONBLOCK_BIT),
 
 	/* fail rest of links */
 	REQ_F_FAIL		= BIT(REQ_F_FAIL_BIT),
@@ -2686,7 +2689,8 @@ static int io_prep_rw(struct io_kiocb *req, const struct io_uring_sqe *sqe)
 		return ret;
 
 	/* don't allow async punt for O_NONBLOCK or RWF_NOWAIT */
-	if ((kiocb->ki_flags & IOCB_NOWAIT) || (file->f_flags & O_NONBLOCK))
+	if ((kiocb->ki_flags & IOCB_NOWAIT) ||
+	    ((file->f_flags & O_NONBLOCK) && !(req->flags & REQ_F_IGNORE_NONBLOCK)))
 		req->flags |= REQ_F_NOWAIT;
 
 	ioprio = READ_ONCE(sqe->ioprio);
@@ -4709,7 +4713,8 @@ static int io_accept(struct io_kiocb *req, unsigned int issue_flags)
 	unsigned int file_flags = force_nonblock ? O_NONBLOCK : 0;
 	int ret;
 
-	if (req->file->f_flags & O_NONBLOCK)
+	if ((req->file->f_flags & O_NONBLOCK) &&
+	    !(req->flags & REQ_F_IGNORE_NONBLOCK))
 		req->flags |= REQ_F_NOWAIT;
 
 	ret = __sys_accept4_file(req->file, file_flags, accept->addr,
diff --git a/include/uapi/linux/io_uring.h b/include/uapi/linux/io_uring.h
index f1f9ac114b51..582eee6e898b 100644
--- a/include/uapi/linux/io_uring.h
+++ b/include/uapi/linux/io_uring.h
@@ -70,6 +70,7 @@ enum {
 	IOSQE_IO_HARDLINK_BIT,
 	IOSQE_ASYNC_BIT,
 	IOSQE_BUFFER_SELECT_BIT,
+	IOSQE_IGNORE_NONBLOCK_BIT,
 };
 
 /*
@@ -87,6 +88,8 @@ enum {
 #define IOSQE_ASYNC		(1U << IOSQE_ASYNC_BIT)
 /* select buffer from sqe->buf_group */
 #define IOSQE_BUFFER_SELECT	(1U << IOSQE_BUFFER_SELECT_BIT)
+/* ignore if file has O_NONBLOCK set */
+#define IOSQE_IGNORE_NONBLOCK	(1U << IOSQE_IGNORE_NONBLOCK_BIT)
 
 /*
  * io_uring_setup() flags
```

**Comment by axboe** [comment]
Seems to work for me, just tested it with a pipe. Even if O_NONBLOCK is set, as long as IOSQE_IGNORE_NONBLOCK is set, we don't return -EAGAIN but rather arm poll and trigger the read once data is posted to the pipe.

**Comment by YoSTEALTH** [comment]
Funny you mentioned pipe I was just thinking if it would work on pipe as well.

**Comment by axboe** [comment]
Oh, patch is against my 5.14 branch fwiw, for-5.14/io_uring

**Comment by axboe** [comment]
Posted: https://lore.kernel.org/io-uring/20210618191940.68303-2-axboe@kernel.dk/

**Comment by axboe** [comment]
An alternative approach would to make this behavior a per-ring kind of thing, which would probably be sufficient as well and save the sqe flag. @isilence what do you think?

**Comment by ciconia** [comment]
@axboe You talk about doing an arm poll after getting -EAGAIN. Would there be any performance costs to setting IOSQE_IGNORE_NONBLOCK for an fd with O_NONBLOCK vs doing the same op on a blocking fd?

**Comment by axboe** [comment]
There would be no performance cost associated with doing that, as the condition would never trigger. So could be safely done without incurring any extra cost.

**Comment by isilence** [comment]
@ciconia, if a file doesn't implement `->read_iter` (in kernel thing) but only `->read` (and equivalents for writes), then IIRC the only way for the file to know about whether it's nonblocking or not is through O_NONBLOCK. In case of missing `read_iter`, io_uring will punt it to internal thread-workers, which in theory should do blocking but in this situation will be busy looping...

Luckily, most does implement ->read_iter, though I already see ib_uverbs_event_read() not having it and doing
```C
if (flags & O_NONBLOCK) return -EAGAIN
```

Interesting to see how much of them are such and how important they're for us.
Have to note that it has nothing to do with the patch Jens sent, just it's probably better off without O_NONBLOCK in the first place if not needed.

**Comment by axboe** [comment]
Right, my statement is true if ->read_iter() is there. ->read() should really just die, and anything that io_uring uses is problematic without the iter variants of read and write. I've mostly been dealing with those on a one-by-one basis, thankfully most things we care about is already using the iter variants.

**Comment by isilence** [comment]
> An alternative approach would to make this behavior a per-ring kind of thing, which would probably be sufficient as well and save the sqe flag. @isilence what do you think?

I'm not picky on implementation side, but saving a bit sounds more alluring. Also, will save applications from keep setting it for each request, especially considering them potentially keeping compatibility with older kernels. Like
```C
if (kernel_is_514())
    sqe->flags |= THE_FLAG.
```

**Comment by axboe** [comment]
Yep, it'd mean that it's relegated to setup time only, which is always preferable to hot path.

**Comment by ciconia** [comment]
> Also, will save applications from keep setting it for each request, especially considering them potentially keeping compatibility with older kernels.

Yeah that's actually the only concern I have. Setting a per-ring flag might be a better solution.

**Comment by ciconia** [comment]
> Also, will save applications from keep setting it for each request, especially considering them potentially keeping compatibility with older kernels.

Yeah that's actually the only concern I have. Setting a per-ring flag might be a better solution.

**Comment by axboe** [comment]
Then it looks like this:

```
commit 85d5f0509fddeb9ee3f446426280a9a4ab28d243
Author: Jens Axboe <axboe@kernel.dk>
Date:   Fri Jun 18 13:15:36 2021 -0600

    io_uring: add IORING_SETUP_IGNORE_ONONBLOCK flag
    
    If a file has O_NONBLOCK set, then io_uring will not arm poll and wait
    for data/space if none is available. Instead, -EAGAIN is returned if
    an IO attempt is made on the file descriptor.
    
    For library use cases, the library may not be in full control of the
    file descriptor, and hence cannot modify file flags through fcntl(2).
    Or even if it can, it'd be inefficient and require 3 system calls to
    check, set, and re-set.
    
    If a ring is setup with IORING_SETUP_IGNORE_ONONBLOCK, that tells io_uring
    to ignore O_NONBLOCK and arm poll to wait for data/space instead, just
    like we would have if O_NONBLOCK wasn't set on the file descriptor.
    
    Suggested-by: Samuel Williams <samuel@codeotaku.com>
    Signed-off-by: Jens Axboe <axboe@kernel.dk>

diff --git a/fs/io_uring.c b/fs/io_uring.c
index fc8637f591a6..214b8cd297cf 100644
--- a/fs/io_uring.c
+++ b/fs/io_uring.c
@@ -2686,7 +2686,8 @@ static int io_prep_rw(struct io_kiocb *req, const struct io_uring_sqe *sqe)
 		return ret;
 
 	/* don't allow async punt for O_NONBLOCK or RWF_NOWAIT */
-	if ((kiocb->ki_flags & IOCB_NOWAIT) || (file->f_flags & O_NONBLOCK))
+	if ((kiocb->ki_flags & IOCB_NOWAIT) ||
+	    ((file->f_flags & O_NONBLOCK) && !(ctx->flags & IORING_SETUP_IGNORE_ONONBLOCK)))
 		req->flags |= REQ_F_NOWAIT;
 
 	ioprio = READ_ONCE(sqe->ioprio);
@@ -4709,7 +4710,8 @@ static int io_accept(struct io_kiocb *req, unsigned int issue_flags)
 	unsigned int file_flags = force_nonblock ? O_NONBLOCK : 0;
 	int ret;
 
-	if (req->file->f_flags & O_NONBLOCK)
+	if ((req->file->f_flags & O_NONBLOCK) &&
+	    !(req->ctx->flags & IORING_SETUP_IGNORE_ONONBLOCK))
 		req->flags |= REQ_F_NOWAIT;
 
 	ret = __sys_accept4_file(req->file, file_flags, accept->addr,
@@ -9755,7 +9757,7 @@ static long io_uring_setup(u32 entries, struct io_uring_params __user *params)
 	if (p.flags & ~(IORING_SETUP_IOPOLL | IORING_SETUP_SQPOLL |
 			IORING_SETUP_SQ_AFF | IORING_SETUP_CQSIZE |
 			IORING_SETUP_CLAMP | IORING_SETUP_ATTACH_WQ |
-			IORING_SETUP_R_DISABLED))
+			IORING_SETUP_R_DISABLED | IORING_SETUP_IGNORE_ONONBLOCK))
 		return -EINVAL;
 
 	return  io_uring_create(entries, &p, params);
diff --git a/include/uapi/linux/io_uring.h b/include/uapi/linux/io_uring.h
index f1f9ac114b51..972fa742119b 100644
--- a/include/uapi/linux/io_uring.h
+++ b/include/uapi/linux/io_uring.h
@@ -98,6 +98,7 @@ enum {
 #define IORING_SETUP_CLAMP	(1U << 4)	/* clamp SQ/CQ ring sizes */
 #define IORING_SETUP_ATTACH_WQ	(1U << 5)	/* attach to existing wq */
 #define IORING_SETUP_R_DISABLED	(1U << 6)	/* start with ring disabled */
+#define IORING_SETUP_IGNORE_ONONBLOCK	(1U << 7)	/* ignore O_NONBLOCK */
 
 enum {
 	IORING_OP_NOP,
```

**Comment by isilence** [comment]
> Pavel, outside of the patch being a quick'n dirty, the intent is _only_ to ignore O_NONBLOCK if set. We should not be fiddling with RWF_NOWAIT or ditto send/recvmsg flags. Just for per-file state, not per request.

I get all the WIP thing, it just seems to me that instead of good understandable model, we solve problems by piling up more complexity on top. Even thinking how to describe the whole situation... It will be a normal async execution, unless there is O_NONBLOCK/RWF_NOWAIT/etc, which would make them nonblocking... Well unless it's specifically O_NONBLOCK and we set IOSQE_IGNORE_NONBLOCK, which would be back normally async. It's kind of a messy model IMHO.

So, there are questions I have:
1) Why can't the userspace embrace O_NONBLOCK, not try to workaround it, or require it to not be set. What are particular cases?

2) Does io_uring even need to honor O_NONBLOCK? Is there any user caring about it? Because there were always cases where it could punt it somewhere and do not return back even if there is no data. There is IOSQE_ASYNC that would do exactly that. And remember I unified RWF_NOWAIT and O_NONBLOCK io_uring specific handling 5.12 or 5.13, so there should again have been some inconsistency.

**Comment by isilence** [comment]
fwiw, I'm not arguing about performance, it's hidden well enough.

**Comment by ciconia** [comment]
> Why can't the userspace embrace O_NONBLOCK, not try to workaround it, or require it to not be set. What are particular cases?

Dealing with EAGAIN in userspace *will* have a performance cost. The specific use case here which lead to @ioquatix opening this issue is the fact that in Ruby 3.0 sockets are non-blocking: https://bugs.ruby-lang.org/issues/17664.

**Comment by axboe** [comment]
The problem with the library situation is that it isn't your fd. Someone else could be reading/writing to it as well, or doing whatever things. So even if doing fcntl get+set, op, then set was viable (that's 3 syscalls...), it's not necessarily a great option.

We can't change how we treat O_NONBLOCK in io_uring at this point without having it be an opt-in kind of thing. This particular use case wants us to ignore it, I'd be willing to bet that there are users right now that depends on O_NONBLOCK returning an -EAGAIN cqe for such a file descriptor. That's totally valid and how we currently behave, and even if you could argue it's a broken use case (it's not), we still couldn't change it now.

**Comment by isilence** [comment]
btw,

```C
static ssize_t sock_read_iter(struct kiocb *iocb, struct iov_iter *to)
{
	...
	if (file->f_flags & O_NONBLOCK || (iocb->ki_flags & IOCB_NOWAIT))
		msg.msg_flags = MSG_DONTWAIT;
	...
}
```

**Comment by isilence** [comment]
> The problem with the library situation is that it isn't your fd. Someone else could be reading/writing to it as well, or doing whatever things. So even if doing fcntl get+set, op, then set was viable (that's 3 syscalls...), it's not necessarily a great option.

The argument is that libraries may have requirements, but I get it.
It just sounds to me that Ruby did something strange and now we're trying to patch it up on the kernel side.

> We can't change how we treat O_NONBLOCK in io_uring at this point without having it be an opt-in kind of thing. This particular use case wants us to ignore it, I'd be willing to bet that there are users right now that depends on O_NONBLOCK returning an -EAGAIN cqe for such a file descriptor. That's totally valid and how we currently behave, and even if you could argue it's a broken use case (it's not), we still couldn't change it now.

I mean that io_uring haven't been providing a consistent model for those, considering apoll and looping in io-wq

**Comment by isilence** [comment]
Even more... that doesn't look good in any case.

```C
static int ____sys_recvmsg(struct socket *sock, struct msghdr *msg_sys,
			   struct user_msghdr __user *msg,
			   struct sockaddr __user *uaddr,
			   unsigned int flags, int nosec)
{
        ...
	if (sock->file->f_flags & O_NONBLOCK)
		flags |= MSG_DONTWAIT;
        ...
}
```

**Comment by YoSTEALTH** [comment]
Every language is going to have their own way of implementing socket and how it handles blocking/non-blocking. In Python you can't even get access to raw socket without it being an object and non-blocking is set using timeout (rabbit hole). I am sure it makes sense in its own ego system. When you try and mix all these into `io_uring`, you need to jump through many hoops.

By design default behavior being asynchronous in io_uring, most people are already confused they still think default is blocking and you need to set `O_NONBLOCK` to make it asynchronous.

Why does io_uring has to worry about `fd` that was created somewhere else? Its going to be a mess trying to support every flag out there and fcntl rules set on it. io_uring has its own set of opens now, let people use that.

It would be better for io_uring to implement a proper `socket` and set its best solution to handle IO properly. This way anyone that uses io_uring will adapt to it.

A small change in io_uring means lots of code/api/design change in higher level language/software.

Everyone knows io_uring is cutting edge, so stick with cutting-edge solution.

**Comment by isilence** [comment]
> Every language is going to have their own way of implementing socket and how it handles blocking/non-blocking. In Python you can't even get access to raw socket without it being an object and non-blocking is set using timeout (rabbit hole). I am

At least a binding to a custom C function doing `fcntl` should do. IIRC it can't block, so have no clue why it needs timeout. For other `fcntl` ops, perhaps.

>  sure it makes sense in its own ego system. When you try and mix all these into `io_uring`, you need to jump through many hoops.

"ego system", nice pun!

But it sounds weird to me, I'd expect any sane language providing access to such internals and allow easy C binding for low-level libraries. Or is it a library vs lang core thing?

> Why does io_uring has to worry about `fd` that was created somewhere else? Its going to be a mess trying to support every flag out there and fcntl rules set on it. io_uring has its own set of opens now, let people use that.

Do you mean mean why io_uring should honor O_NONBLOCK and other flags? Because they are already supported, no more no less. If compatibility wasn't a problem, I'd much prefer to leave IOSQE_NONBLOCK and only it controlling the behaviour, but can't be done.  

> It would be better for io_uring to implement a proper `socket` and set its best solution to handle IO properly. This way anyone that uses io_uring will adapt to it.

What do you mean by "a proper socket"? socket(2)? The whole net stack?
 
> A small change in io_uring means lots of code/api/design change in higher level language/software.
> 
> Everyone knows io_uring is cutting edge, so stick with cutting-edge solution.

What do you mean? who should or going to stick with it?

**Comment by YoSTEALTH** [comment]
First of all I been up for way too long, so I am running on fumes here.

> At least a binding to a custom C function doing fcntl should do. IIRC it can't block, so have no clue why it needs timeout. For other fcntl ops, perhaps.

They do use fcntl and such, I am sure its more to do with backwards compatibility and or bugs. Maybe they are mixing `O_NONBLOCK` + timeout into one. https://github.com/python/cpython/blob/main/Modules/socketmodule.c

> "ego system", nice pun!

I meant to say ecosystem :)

> But it sounds weird to me, I'd expect any sane language providing access to such internals and allow easy C binding for low-level libraries. Or is it a library vs lang core thing?

If makes using Python simple and nice!

For example when `io_uring_prep_close` was added I kept trying to close socket `fd` that was created in Python and after a week or two of fun times! (new pun) I found out that even though `fd` was being closed python class would keep that `fd` open. So the solution is to detach it from the class then close. Which someone did implemented and most likely hardly anyone used it. So a "solution" does exist!

Who knows what else is being attached to that socket `fd`. Like I have to use `recv` with `MSG_DONTWAIT` flag or else socket goes into stuck mode. I haven't had the time to figure out whats actually happening with it. Monkey patch works for now.

> Do you mean mean why io_uring should honor O_NONBLOCK and other flags? Because they are already supported, no more no less. If compatibility wasn't a problem, I'd much prefer to leave IOSQE_NONBLOCK and only it controlling the behaviour, but can't be done.

It makes no sense for liburing to support `O_NONBLOCK` (maybe io_uring can use it for internal use). Its like someone attaching a piece of paper to your back saying "Kick Me".

> What do you mean by "a proper socket"? socket(2)? The whole net stack?

If io_uring implements `io_uring_prep_socket`+ everyone will start using it is what I mean. Like how there is `io_uring_prep_openat*` and its such an awesome feature.

> What do you mean? who should or going to stick with it?

Look at this for example https://bugs.python.org/issue4761 someone implemented `openat*` nearly 10 years ago and even today there is no `openat*` in Python stdlib. Its what bad call mixed with compatibility mess looks like. Everyone lost out on great features cause of it. I am saying lets (io_uring) not follow that patch of trying to provide backwards compatibility and look at next 1-20+ years mark where io_uring is actually going to be used heavily.

Once you have basic IO features implemented then you can worry about keeping it compatible, not when its still in early development and trying to support feature that isn't even being used.

**Comment by ioquatix** [comment]
Thanks for all the great discussion.

There are several places where we cannot create our IO. Opaque libraries like `libmariadb` which internally create sockets and have some (crappy) interfaces for getting the file descriptor out for event driven I/O is one example.

Therefore, trying to funnel people into a specific way to construct file descriptors won't solve this problem generally, although I understand why that might be a nice approach.

For Ruby, it's my decision to implement non-blocking I/O generally. In Ruby, when we perform I/O operation, like `read`, and it returns `EAGAIN` we automatically enter event loop. When I/O is blocking, this won't happen and the thread will block. Both are okay, but we wanted a ubiquitous way to implement event driven concurrency, i.e. `read` -> `EAGAIN` -> `wait_readable` -> scheduler -> `select`/`epoll`/`kqueue`/`uring`.

But we recently introduced a hook which allows some operations like `read` to go directly to the event loop. This is where the problem happens, because our I/O is non-blocking, it can go like this: `read` -> scheduler -> `uring`.

Within the same process, or even the same file descriptor, in theory it can go through both paths, and even now, the current implementation should handle `uring` resulting in `EAGAIN` for compatibility.

But I think that the most common use case is going to be expecting `uring` to perform the read and don't expect `EAGAIN` result. As discussed, there is a user space cost to this.

Personally, the global per-uring flag makes sense to me from the place I'm coming from. I believe it means we can effectively implement "blocking" `read` as outlined above via the event loop and `IORING_OP_READV`.

**Comment by ioquatix** [comment]
By the way, I'm slowly rewriting Ruby's IO layer to be event driven at heart, and we will take advantage and look to the interface design of `io_uring` to guide us. Most Ruby applications are hosted on Linux so this is the number 1 platform in terms of where we want to spend our effort w.r.t performance and concurrency.

**Comment by ciconia** [comment]
> It just sounds to me that Ruby did something strange and now we're trying to patch it up on the kernel side.

Adding this flag, whether it's per-request or per-ring, would help tremendously in integrating io_uring in a whole bunch of "ecosystems" that have established APIs and need to maintain backward-compatibility. *And* it's opt-in as @axboe said.

**Comment by lano1106** [comment]
I just want to chime in that this is a nice addition to io_uring. I got confronted with this problem as well. By definition, as soon as you start using an event polling loop framework, all fds are going to be nonblocking.

In my application, I was, fortunately, able to identify where to remove the O_NONBLOCK flag to respect the existing nonblocking connection semantic while transitioning to blocking io_uring async i/o. It was where the openssl context was created!

**Comment by YoSTEALTH** [comment]
`*IGNORE_NONBLOCK` per `io_uring` sounds like a better idea then per `sqe` based.

**Comment by isilence** [comment]
I get that it's useful for libraries, my problem is that it looks like a quick fix. And apparently this gut feeling is not in vain, see `sock_read_iter` and `____sys_recvmsg`, if I haven't missed something, socket read and recvmsg (most probably + sendmsg and write) requests won't work well in any case. In cases when they can't serve data immediately and require blocking It will be doing ugly busy looping (see `io_wq_submit_work()`) that might hurt performance. Those were spotted accidentally, there may be others.

In the ideal world, I'd love to see Ruby fixing the regression (was it a recent change?), iow drop O_NONBLOCK and start using RWF_NOWAIT/MSG_DONTWAIT and so on. But I admit that there may be complications and implications, e.g. because of files not having `reat_iter`.

**Comment by isilence** [comment]
> > Do you mean mean why io_uring should honor O_NONBLOCK and other flags? Because they are already supported, no more no less. If compatibility wasn't a problem, I'd much prefer to leave IOSQE_NONBLOCK and only it controlling the behaviour, but can't be done.
> 
> It makes no sense for liburing to support `O_NONBLOCK` (maybe io_uring can use it for internal use). Its like someone attaching a piece of paper to your back saying "Kick Me".

Agree, but backward compatibility. See Jens answered above. Also, some(?) files/sockets/etc honor it already, so it's not controlled by io_uring and is a deeper problem.

> > What do you mean by "a proper socket"? socket(2)? The whole net stack?
> 
> If io_uring implements `io_uring_prep_socket`+ everyone will start using it is what I mean. Like how there is `io_uring_prep_openat*` and its such an awesome feature.

I need to ask, why it is needed? comparing to calling a `socket(2)`
1) Does it add something new that couldn't have been done before?
2) How time creating sockets compares to everything else? I.e. time spent on I/O and computations.
3) I guess openat can block, e.g. like with net filesystems. Does socket(2)?

Can be added if there is a good justification for that

**Comment by YoSTEALTH** [comment]
> I need to ask, why it is needed? comparing to calling a socket(2)

Compatibility, standardization, Usability, Trust ...

It literally takes me < 5 minutes to write a wrapper around `socket`. Should I really provide that to users? Today the answer might be yes, and tomorrow if io_uring implemented its own version of `socket` it will be breaking compatibility.

Just because I implemented something into my project does that mean other language users gets to use it? No. Now its not standardized. Everyone is inventing their own version of the wheel.

Things you could do with `io_uring_prep_socket` you can't do with normal `socket` (without high modification), like linking, timeout, ... there are new innovation waiting to happen.

Most likely people wont trust a wana be async socket if its not coming from io_uring (at the very least).

This is where the problem starts.

You end-up using next best thing users will trust like Python, Ruby's socket module and run into weird problems.

Dynamic language (in most cases) are designed to not break (vs something like C where you can break anything). Things like open, socket are prized features and strictly controlled. They have too many checks and fallback.

If you are saying "io_uring is definitely not "just" another event handling interface."~Jens then you need to include few batteries to go along with the library.

**Comment by ioquatix** [comment]
@isilence are you saying that it would be better for sockets to use poll (via io_uring) and local files to use blocking read (via io_uring)?

**Comment by isilence** [comment]
> > I need to ask, why it is needed? comparing to calling a socket(2)
> 
> Compatibility, standardization, Usability, Trust ...
> 
> It literally takes me < 5 minutes to write a wrapper around `socket`. Should I really provide that to users? Today the answer might be yes, and tomorrow if io_uring implemented its own version of `socket` it will be breaking compatibility.
> 
> Just because I implemented something into my project does that mean other language users gets to use it? No. Now its not standardized. Everyone is inventing their own version of the wheel.

Seems, there is misunderstanding. By `io_uring_prep_socket()` do you mean implement `socket(2)`?
This https://www.man7.org/linux/man-pages/man2/socket.2.html

Both, io_uring new opcode and `socket(2)` would return back the same socket with exactly same internals, and either would be compatible with both io_uring (e.g. issuing requests with it) and all standard syscalls like send(2), read(2), ioctl(2) and so on.  If you mean implementing internally your own socket, like adding a new socket family, reimplementing net stack (e.g. creating your own tcp or whatever), it's just not needed.

That's putting aside discussion about your library design and whether it's better to extract an fd from already existing Python objects and so on.

> Things you could do with `io_uring_prep_socket` you can't do with normal `socket` (without high modification), like linking, timeout, ... there are new innovation waiting to happen.

Timeout `socket(2)`? Look, it's very fast and not blocking AFAIK, correct me if someone knows otherwise. The heavy blocking part is connect/accept, which io_uring already supports.

> Dynamic language (in most cases) are designed to not break (vs something like C where you can break anything). Things like open, socket are prized features and strictly controlled. They have too many checks and fallback.

Emm? Break compatibility or at runtime (e.g. segfault)?

> If you are saying "io_uring is definitely not "just" another event handling interface."~Jens then you need to include few batteries to go along with the library.

Ok, which ones? Ultimately, it gets what it needs to get

**Comment by isilence** [comment]
> @isilence are you saying that it would be better for sockets to use poll (via io_uring) and local files to use blocking read (via io_uring)?

It may be, but that's not the point.

I'd rather say that the problem should be solved fundamentally. One way is to not set O_NONBLOCK in the userspace in the first place. Either to _somehow_ make the kernel internals ignore O_NONBLOCK, and that has nothing to do with io_uring directly. At least for devices and files of interest like sockets. Also just found O_NONBLOCK honored in pipe' `->*_iter()` callbacks.

**Comment by isilence** [comment]
> @isilence You are not providing `socket` because of highly technical reason but more as a supporting feature (prevent people from making bad choices).
> 
> Think of io_uring as if you are buying a car. You get the best engine, suspension, frame, but you don't get a seat with io_uring. And you are telling me its just a seat! you can buy that anywhere, use a stool. Get what I mean? Do you really want to drive around in a best car you can get with a stool as a seat? With no seat belt, not fashioned to the frame?

And together with a car you additionally get a nice fruit blender 3000 with a book of ice cream recipes. That's a problem with such metaphors...

**Comment by YoSTEALTH** [comment]
> The bottom line, we're always open to suggestions and requests, but I'd expect some justification.

O, I thought it was obvious. The main reason this topic exists is since `io_uring` doesn't have `io_uring_prep_socket` @ioquatix had to use the next best thing, which is to use Ruby's `socket` module https://bugs.ruby-lang.org/issues/17664 and since Ruby's 3.0 `socket` releases with `O_NONBLOCK` as its default flag, now ioquatix is stuck with an `fd` that has `O_NONBLOCK` on it thus causing problems with `io_uring`.

Sure @axobe patch will help with ioquatix issue, but what about the next languages socket problem?

This is not just a Ruby issues, this is major problem in all dynamic language, I thought I made that very clear. What more justification is needed?

**Comment by victorstewart** [comment]
@YoSTEALTH 

> since io_uring doesn't have io_uring_prep_socket @ioquatix had to use the next best thing, which is to use Ruby's socket module

but if the socket can be created outside the Ruby environment, and as you suggest by an io_uring operation, then there is no difference than right now using the `socket(2)` interface to create the exact same socket object that such an io_uring operation would produce.

the problem OP is calling attention to is when forced to work with a socket that you do not / can not create.

**Comment by isilence** [comment]
> > The bottom line, we're always open to suggestions and requests, but I'd expect some justification.
> 
> O, I thought it was obvious. The main reason this topic exists is since `io_uring` doesn't have `io_uring_prep_socket` 

I just want to emphasise that it might be useful, but I don't see how yet. 

> @ioquatix had to use the next best thing, which is to use Ruby's `socket` module https://bugs.ruby-lang.org/issues/17664 and since Ruby's 3.0 `socket` releases with `O_NONBLOCK` as its default flag, now ioquatix is stuck with an `fd` that has `O_NONBLOCK` on it thus causing problems with `io_uring`.
> 
> Sure @axobe patch will help with ioquatix issue, but what about the next languages socket problem?

The problem is that it will not solve the problem completely, iow it will get rid of -EAGAIN but performance still will be suboptimal. 
 
> This is not just a Ruby issues, this is major problem in all dynamic language, I thought I made that very clear. What more justification is needed?

Are you about `io_uring_prep_socket`? Because it doesn't matter whether you create a socket via io_uring or with `socket(2)`. The problem is lying further, 1) either it's wrapped into a new separate class, and so it's not compatible with all the libraries but can be finely controlled, or 2) you do create a standard Python/Ruby/etc. socket class or so, and then you should follow the rules and set O_NONBLOCK for in case of the current Ruby situation, otherwise someone will use it with Ruby's sendmsg or something, it will unexpectedly block and that would be a bug.

Just to be clear, having or not `io_uring_prep_socket` is _completely_ orthogonal to the non-block issue. See also a @victorstewart 's reply above saying the same thing, if he put it into words more clearly.

**Comment by ioquatix** [comment]
> The problem is that it will not solve the problem completely, iow it will get rid of -EAGAIN but performance still will be suboptimal.

This is the part I would like to understand in more detail.

Are you able to guide me on where this "suboptimal" behaviour comes from, and what does "optimal" look like?

**Comment by isilence** [comment]
> > The problem is that it will not solve the problem completely, iow it will get rid of -EAGAIN but performance still will be suboptimal.
> 
> This is the part I would like to understand in more detail.
> 
> Are you able to guide me on where this "suboptimal" behaviour comes from, and what does "optimal" look like?

io_uring first tries to execute requests in a nonblocking fashion to see if the data is already there, that's the fast path and it's ok. Then it can try polling and other tricks, but eventually it _might_ end up in an internal IO worker (aka io-wq). io-wq, would try to execute it blocking with:

```C
while (io_send(BLOCKING) == -EAGAIN)
	cond_resched(); // simplifying, in-kernel sleep(0);
```

It would think that it's blocking and looping is not a problem. However, if the net stack sees O_NONBLOCK and somewhere internally sets MSG_DONTWAIT, as indicated by posts before, it will be looping burning CPU time. Simplified and cutting details, but close enough.

For when it can end up in io-wq, I'd expect it happening often enough for QD>1. Not sure without checking the net code whether QD includes all sockets of a NIC or only one, but that's _kind of_ #QD parallel `poll(2)`s racing for data.
@lano1106 was recently interested why and how it happens. Any chance you have some data left on how many requests end up in io-wq for your case?

**Comment by isilence** [comment]
I'm wrong about looping though, if there is still no data by the time a request gets into io-wq, it will be failed with -EAGAIN.

**Comment by YoSTEALTH** [comment]
@victorstewart @isilence we can talk more about socket issue at https://github.com/axboe/liburing/issues/234

**Comment by YoSTEALTH** [comment]
FYI, I am finding that if I disallow using `O_NONBLOCK`, I also have to disallow `RWF_NOWAIT`. There doesn't seem a point in allowing `RWF_NOWAIT` then catching `EAGAIN` just to poll it again at user-space, while default `io_uring` already does a better job at it.

Rater then `io_uring` implementing a complex solution it might be better for developer to do: `fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) & ~O_NONBLOCK);` this is a user-space problem let developer deal with it.

Also for Liburing to discourage using `O_NONBLOCK`, `RWF_NOWAIT`, ... sure these were awesome feature back in the day, but what really is the advantage in using those with io_uring?

**Comment by ioquatix** [comment]
@YoSTEALTH

1. epoll, kqueue need `O_NONBLOCK`, as in: try to perform `read` -> `EAGAIN` -> `epoll`.
2. uring doesn't need `O_NONBLOCK` as in: `IORING_OP_READV`.
3. Both implementations can exist at the same time in a programming language, e.g.

```
epoll = Event::Backend::EPoll.new
uring = Event::Backend::URing.new
```

In order to make this work, when calling `io.read`, we simply perform `read` system call. If it returns EAGAIN, we punt it off to the event loop. However, we also have the ability to completely replace `read` system call with the event backend's own implementation. This is what we are experimenting with for `uring` backend, i.e. using `IORING_OP_READV`.

In terms of our decision to make IO nonblocking: Either IO is `O_NONBLOCK` and this works fine for `epoll`, but cause some efficiency issues in `uring`. If we don't mark IO as non-block, we need to give flags to `read` operation when invoking it from `epoll` backend, so that it would return `EAGAIN` some how. If that was possible in all the cases we need it, across all the platforms we support, it would be an option to consider but I'm not sure it is.

We can certainly have the `uring` implementation handle EAGAIN too, but this just seems inefficient. We will have to do it anyway for backwards compatibility.

**Comment by YoSTEALTH** [comment]
Sure, you can have both implementation side by side. Are you sharing `fd` between `epoll` and `io_uring`? That seems odd. That said, wouldn't it be easier if you just made an duplicate of `fd`? https://man7.org/linux/man-pages/man2/dup.2.html

This way you can set `io_uring` `fd` with `O_NONBLOCK` and it shouldn't effect your `epoll`.

**Comment by YoSTEALTH** [comment]
> when calling io.read, we simply perform read system call. If it returns EAGAIN, we punt it off to the event loop. 

O..., I think I get what you are doing... You simply take the initial `socket` and try to `socket.recv()` and only call poll if it blocks, right? I know that's how it works in old epoll world.

> We can certainly have the uring implementation handle EAGAIN too, but this just seems inefficient. We will have to do it anyway for backwards compatibility. 

Yes, leave `O_NONBLOCK` as is and catch `EAGAIN` and poll it using `io_uring_prep_poll_add`.

If you are worried about efficiency, don't! You have already lost efficiency when you call `socket.recv()` directly and `EAGAIN` is not triggered as often as you think it does.

Secondly, I get this weird feeling like if you tell io_uring to ignore `O_NONBLOCK` it might block your Ruby's event-manager for a long time/till external IO event happens. It depends how you are handling `*wait*` state.

**Comment by ioquatix** [comment]
> Secondly, I get this weird feeling like if you tell io_uring to ignore O_NONBLOCK it might block your Ruby's event-manager for a long time/till external IO event happens. It depends how you are handling *wait* state.

As long as nothing else is happening, that's fine. It's the heart of the event loop. If some other work is scheduled, we set a timeout (which could be 0).

**Comment by ioquatix** [comment]
I've been playing around with this more. I implemented support both blocking and non-blocking by punting the read/write operation to `io_wait` if it returns `EAGAIN`. That means that in theory, we can take advantage of this flag in the future without any compatibility concerns, but I assume simply better performance since more `io_read`/`io_write` will be fully serviced by the kernel rather than going to poll.

**Comment by ioquatix** [comment]
@axboe was this merged?

https://lore.kernel.org/io-uring/20210618221040.91075-2-axboe@kernel.dk/

I'm running `5.19.9-arch1-1` but I don't see it.

**Comment by ioquatix** [comment]
If it's not merged, might I suggest a slightly more readable name? e.g. `IORING_SETUP_IGNORE_NONBLOCK`.

**Comment by ioquatix** [comment]
@axboe any update?

**Comment by mahkoh** [comment]
I've encountered this problem in my wayland compositor. It used a mix of classical apis (accept4, connect, sendmsg, recvmsg) and io_uring (sendmsg, recvmsg). All sockets were created with SOCK_NONBLOCK to accommodate the classical apis. The code using io_uring did not handle EAGAIN and instead treated any error as fatal. Everything would run fine for several hours but then a client (usually chrome) would get killed because sendmsg or recvmsg (via io_uring) returned EAGAIN. Like I said, this only happened very rarely, not more than once or twice per day.

The solution for me was to replace all classical apis by io_uring and make the sockets blocking.

However, I wonder if this might be a bug in the io_uring implementations of sendmsg or recvmsg. I've read the code of the sendmsg implementation and my understanding is as follows:

- When sendmsg is issued the first time, io_uring will force non-blocking behavior (MSG_NOWAIT). It doesn't matter whether the socket is not-blocking or not.
- If this returns EAGAIN, a poll mechanism is armed and io_uring waits for the socket to become writable.
- Once the poll mechanism signals readiness, io_uring issues the the sendmsg again, this time not forcing non-blocking behavior.

As I understand it, a sendmsg submitted via io_uring can only return EAGAIN in the last step. However, I don't see how this can happen. I know for a fact that there is only a single writer writing to this socket (the one via io_uring). Is it possible that the second step can cause spurious wakeups even if the socket is not writable?

In any case, the behavior of sendmsg via io_uring with non-blocking sockets seems quite irregular and unpredictable. Usually it will block (i.e. the submitted work queue item does not complete) but very rarely it will spuriously fail early with EAGAIN. This does not seem useful whether whether you want non-blocking behavior or not.

**Comment by ioquatix** [comment]
@axboe any chance to get some attention here? Thanks!

**Comment by axboe** [comment]
I don't have any updates to share. Rather than make this a per-sqe kind of flag, may make more sense to have it as a ring setup (or register enable/disable) flag instead - if we get -EAGAIN and the file is nonblocking, treat it like it wasn't and still go through the poll mechanism. If the file isn't pollable, we'd still end the sqe with -EAGAIN as there's nothing sane we can do at that point.

**Comment by axboe** [comment]
Wrote a quick test for this, and on the current kernels, I do see poll getting armed just fine for a O_NONBLOCK socket. Now I'm curious if we inadvertently "fixed" this, or...? Works regardless of whether or not read or recv is used, here's my output:

```
cqe 2: res=32
cqe 1: res=32
```

I'll attach the test app in the next comment.

**Comment by axboe** [comment]
```
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <liburing.h>

int main(int argc, char *argv[])
{
	struct io_uring ring;
	int ret, i;
	int fds[2];
	struct io_uring_cqe *cqe;
	struct io_uring_sqe *sqe;
	char src[32], dst[32];
	int flags;

	io_uring_queue_init(8, &ring, 0);

	if (socketpair(PF_LOCAL, SOCK_STREAM, 0, fds) < 0) {
		perror("socketpair");
		return 1;
	}

	flags = fcntl(fds[0], F_GETFL, 0);
	flags |= O_NONBLOCK;
	fcntl(fds[0], F_SETFL, flags);

	sprintf(src, "This is the source buffer\n");
	
	sqe = io_uring_get_sqe(&ring);
#if 1
	io_uring_prep_recv(sqe, fds[0], dst, sizeof(dst), 0);
#else
	io_uring_prep_read(sqe, fds[0], dst, sizeof(dst), 0);
#endif
	sqe->user_data = 1;

	ret = io_uring_submit(&ring);
	if (ret != 1)
		return ret;

	usleep(100000);

	sqe = io_uring_get_sqe(&ring);
	io_uring_prep_send(sqe, fds[1], src, sizeof(src), 0);
	sqe->user_data = 2;

	ret = io_uring_submit(&ring);
	if (ret != 1)
		return ret;

	for (i = 0; i < 2; i++) {
		ret = io_uring_wait_cqe(&ring, &cqe);
		if (ret) {
			fprintf(stderr, "wait ret %d\n", ret);
			return 1;
		}
		printf("cqe %d: res=%d\n", (int) cqe->user_data, cqe->res);
		io_uring_cqe_seen(&ring, cqe);
	}

	io_uring_queue_exit(&ring);
	close(fds[0]);
	close(fds[1]);
	return 0;
}
```

**Comment by mahkoh** [comment]
>Wrote a quick test for this, and on the current kernels, I do see poll getting armed just fine for a O_NONBLOCK socket. Now I'm curious if we inadvertently "fixed" this, or...?

In my case I only encountered this issue at most once per day with dozens of O_NONBLOCK sockets being used concurrently.

**Comment by axboe** [comment]
OK, so I think that may be because you end up arming poll, and then it triggers but not ready yet, and at that point it'd get punted to io-wq and errored with -EAGAIN. This can happen with spurious poll wakeups. I realize this can be hard to test if it only happens once per day, but something like this would most likely fix that case:

```
diff --git a/io_uring/poll.c b/io_uring/poll.c
index 8339a92b4510..371ad161ae6d 100644
--- a/io_uring/poll.c
+++ b/io_uring/poll.c
@@ -694,8 +694,10 @@ int io_arm_poll_handler(struct io_kiocb *req, unsigned issue_flags)
 		return IO_APOLL_ABORTED;
 	if (!file_can_poll(req->file))
 		return IO_APOLL_ABORTED;
+#if 0
 	if ((req->flags & (REQ_F_POLLED|REQ_F_PARTIAL_IO)) == REQ_F_POLLED)
 		return IO_APOLL_ABORTED;
+#endif
 	if (!(req->flags & REQ_F_APOLL_MULTISHOT))
 		mask |= EPOLLONESHOT;
```

where basically we currently say "if poll has already been armed and we didn't get any data, don't arm it again" to avoid a potential infinite loop due to a buggy driver/file that keeps saying we can read/write yet we get -EAGAIN from it.

**Comment by mahkoh** [comment]
I remember that when I looked at this at the time I also assumed that this was related to spurious wakeups and that removing that line would fix it. I don't remember the exact context of that code but ever since I removed O_NONBLOCK from all of my sockets, the problem has gone away. Is this code not hit if the socket is blocking?

**Comment by axboe** [comment]
If the socket is blocking, then the io-wq helper thread would sit and wait for data/space. So it wouldn't be optimal, but it's a safer punt.

But I do think we should improve this re-poll behavior. As long as we can detect if we're entering a poll loop and abort at that point, there's no reason not to allow indefinite retries. I'll take a look at that.

**Comment by mahkoh** [comment]
>If the socket is blocking, then the io-wq helper thread would sit and wait for data/space.

Right, that makes sense.

**Comment by axboe** [comment]
Something like this should work Good Enough (tm) in practice I suspect:

https://git.kernel.dk/cgit/linux-block/commit/?h=io_uring-6.3&id=9658ca75fe2dd036a2b23496b5d8d7f0850a5d41

As we only expect this to happen spuriously, we just need some kind of cap on the poll retries. I made it 128, which should be more than plenty.

**Comment by ioquatix** [comment]
Thanks, I do not fully understand the above discussion.

Does this mean `O_NONBLOCK` polls will now generally not return EAGAIN except in some extreme edge cases?

**Comment by axboe** [comment]
That's pretty much what it means, yes.

**Comment by axboe** [comment]
BTW, here's a test case that does detect it. Fails with the current kernels, works with the mentioned patch:

```
/* SPDX-License-Identifier: MIT */
/*
 * Test -EAGAIN handling on repeated poll wakeup without having transferred
 * any data yet.
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>

#include "liburing.h"
#include "helpers.h"

int main(int argc, char *argv[])
{
	struct io_uring_cqe *cqe;
	struct io_uring_sqe *sqe;
	char src[32], dst[32];
	struct io_uring ring;
	int flags, ret, i;
	int fds[2];

	if (argc > 1)
		return T_EXIT_SKIP;

	t_create_ring(8, &ring, 0);

	if (socketpair(PF_LOCAL, SOCK_STREAM | SOCK_NONBLOCK, 0, fds) < 0) {
		perror("socketpair");
		return T_EXIT_FAIL;
	}

	flags = fcntl(fds[0], F_GETFL, 0);
	flags |= O_NONBLOCK;
	fcntl(fds[0], F_SETFL, flags);

	sprintf(src, "This is the source buffer\n");
	
	sqe = io_uring_get_sqe(&ring);
	io_uring_prep_read(sqe, fds[0], dst, sizeof(dst), 0);
	sqe->user_data = 1;

	sqe = io_uring_get_sqe(&ring);
	io_uring_prep_read(sqe, fds[0], dst, sizeof(dst), 0);
	sqe->user_data = 2;

	ret = io_uring_submit(&ring);
	if (ret != 2) {
		fprintf(stderr, "submitted %d of 2\n", ret);
		return T_EXIT_FAIL;
	}

	usleep(100000);

	sqe = io_uring_get_sqe(&ring);
	io_uring_prep_send(sqe, fds[1], src, sizeof(src), 0);
	sqe->user_data = 10;

	ret = io_uring_submit(&ring);
	if (ret != 1) {
		fprintf(stderr, "submitted %d of 1\n", ret);
		return T_EXIT_FAIL;
	}

	/* we should have the first receive and the send */
	for (i = 0; i < 2; i++) {
		ret = io_uring_wait_cqe(&ring, &cqe);
		if (ret) {
			fprintf(stderr, "wait ret %d\n", ret);
			return T_EXIT_FAIL;
		}
		if (cqe->res != sizeof(src)) {
			fprintf(stderr, "bad res: %d\n", cqe->res);
			return T_EXIT_FAIL;
		}
		io_uring_cqe_seen(&ring, cqe);
	}

	/* make sure the last cqe has a chance to post, if any */
	usleep(100000);

	/*
	 * if we find a cqe here, it's probably -EAGAIN. we should not
	 * see any, the second receive should've been rearmed.
	 */
	ret = io_uring_peek_cqe(&ring, &cqe);
	if (!ret) {
		fprintf(stderr, "Unexpected cqe: %d\n", cqe->res);
		return T_EXIT_FAIL;
	}

	io_uring_queue_exit(&ring);
	close(fds[0]);
	close(fds[1]);
	return T_EXIT_PASS;
}
```

**Comment by ciconia** [comment]
How will this work for other types of files: real files, pipes, eventfd etc?

**Comment by axboe** [comment]
Regular files aren't pollable and hence not susceptible to this kind of thing. But apart from that, doesn't matter if it's a pipe/socket/eventfd etc, it'll work the same on either one.

**Comment by ioquatix** [comment]
I will test this on the latest kernel once it's released and close the issue if possible.

**Comment by vasama** [comment]
What's the status of this issue? Is `IORING_SETUP_IGNORE_ONONBLOCK` going anywhere?

**Comment by axboe** [comment]
Should be the case now.

**Comment by ioquatix** [comment]
I will try it out again.

**Comment by ioquatix** [comment]
I've done a basic test, and I'm not seeing EAGAIN any more. However, it's difficult to compare since I don't really have a system to reproduce the old behaviour.
